### PR TITLE
[JENKINS-43610] Split Trilead out from Core

### PR DESCRIPTION
**Pull request by kuisathaverat**
See [JENKINS-43610](https://issues.jenkins-ci.org/browse/JENKINS-43610).

Trilead-ssh2 is bundled in Jenkins core this will allow using trilead-api-plugin instead and remove the Jenkins Core dependence of trilead-ssh2.

It is a dependencies change, so it will be tested by the regular test, ATH, and PCT

Changes on trilead-ssh2 version include in the plugin (trilead-ssh2-build-217-jenkins-16, 2019-03-31):

- [JENKINS-55352] update net.i2p.crypto:eddsa to 0.3.0 
- Connection timeout handler not canceled when an exception occurs during a connection attempt. If the Connection instance is reused for further attempts, this can lead to the disconnection of an active connection.
- [JENKINS-56821] Eliminating hot lock from `TimeoutService`

### Proposed changelog entries

* Major RFE: Split Trilead SSH out from Core, [trilead-ssh2](https://github.com/jenkinsci/trilead-ssh2) library is no longer bundle in Jenkins Core, it is provided as a plugin with [trilead-api-plugin](https://github.com/jenkinsci/trilead-api-plugin) for further details see [JENKINS-43610](https://issues.jenkins-ci.org/browse/JENKINS-43610)

### Submitter checklist

- [X] JIRA issue is well described
- [X] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [X] Appropriate autotests or explanation to why this change has no tests
- [X] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@oleg-nenashev @MarkEWaite @dwnusbaum @daniel-beck @jvz @Wadeck @batmat @olivergondza @jglick

### CODE REVIEW

File: core/src/main/resources/jenkins/split-plugins.txt
Comment (kuisathaverat) [inline core/src/main/resources/jenkins/split-plugins.txt]: Jenkins Core version TBD

File: core/pom.xml
Comment (kuisathaverat) [inline core/pom.xml]: There is a dependency on `jbcrypt` in Jenkins Core so we have to add the dependency

```
/home/jenkins/workspace/Core_jenkins_PR-4085/core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:[68,26] error: package org.mindrot.jbcrypt does not exist
```

**Review by jvz — APPROVED** [review]

File: core/pom.xml
Comment (jglick) [reply core/pom.xml]: FTR, from current `master`:

```
$ mvn -f core dependency:tree
…
[INFO] +- org.jenkins-ci:trilead-ssh2:jar:build-217-jenkins-14:compile
[INFO] |  \- org.connectbot.jbcrypt:jbcrypt:jar:1.0.0:compile
…
```

File: core/src/main/resources/jenkins/split-plugin-cycles.txt
Comment (jglick) [inline core/src/main/resources/jenkins/split-plugin-cycles.txt]: I do not think you understand what this file is for. Revert the changes to this file.

File: core/src/main/resources/jenkins/split-plugins.txt
Comment (jglick) [inline core/src/main/resources/jenkins/split-plugins.txt]: Not specific to Java version AFAIK.

```suggestion
trilead-api 2.181 1.0.3
```

**Review by jglick — CHANGES_REQUESTED** [review]
`split-plugin-cycles.txt` looks incorrect. You also neglected to package the newly detached plugin in `war/pom.xml` (see the section with the rest of them).

Review of https://github.com/jenkinsci/trilead-api-plugin (since there is no PR for to comment on):

* delete `src/main/java/` and the pointless `Plugin`
* POM should not contain `http` URLs; use for example `https://opensource.org/licenses/MIT` and for the `url` follow the redirect to the current location
* `jenkins.version` should for now be the incremental `2.182-rcXXX` from this PR ([guidelines](https://github.com/jenkinsci/incrementals-tools/blob/master/README.md#usage-in-plugin-poms)), to be updated to `2.182` or whatever when released: the lib must not be offered on the UC to users of cores prior to the split
* `maven-javadoc-plugin` config can be deleted, just use a current parent POM
* use `<relativePath />`

**Comment by jglick** [comment]
Somewhere @daniel-beck or @oleg-nenashev captured a video of me talking about split plugins BTW.

**Comment by kuisathaverat** [comment]
suggested changes on trilead-api-plugin https://github.com/jenkinsci/trilead-api-plugin/pull/5

File: core/src/main/resources/jenkins/split-plugin-cycles.txt
Comment (kuisathaverat) [reply core/src/main/resources/jenkins/split-plugin-cycles.txt]: I've reverted the changes

File: core/src/main/resources/jenkins/split-plugins.txt
Comment (kuisathaverat) [reply core/src/main/resources/jenkins/split-plugins.txt]: done, I also update core to 2.182

File: core/src/main/resources/jenkins/split-plugins.txt
Comment (jglick) [inline core/src/main/resources/jenkins/split-plugins.txt]: For now, to make this build, this would have to be an incremental version, as I did for example in #3988.

**Review by jglick — CHANGES_REQUESTED** [review]
Still missing the detached plugin in `war/pom.xml`.

**Comment by kuisathaverat** [comment]
the CI fails to find the trilead-api 1.0.4-rc17.8e3daef99f43 incremental version but it is there https://repo.jenkins-ci.org/incrementals/org/jenkins-ci/plugins/trilead-api/1.0.4-rc17.8e3daef99f43/

**Comment by jglick** [comment]
> the CI fails to find the trilead-api 1.0.4-rc17.8e3daef99f43 incremental version

You need to do [this](https://github.com/jenkinsci/jenkins/pull/3988/files#diff-cb9415d4b4ee6a3281d14807d6ddaa3fR1).

**Review by oleg-nenashev — APPROVED** [review]

File: core/src/main/resources/jenkins/split-plugins.txt
Comment (jglick) [inline core/src/main/resources/jenkins/split-plugins.txt]:
```suggestion
trilead-api 2.184 1.0.4-rc17.8e3daef99f43
```

though it may need to be adjusted immediately before merge.

**Review by jglick — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
Will merge tomorrow if no negative feedback

**Comment by kuisathaverat** [comment]
@oleg-nenashev Should I release trilead-api 1.0.4 now?

**Comment by oleg-nenashev** [comment]
@kuisathaverat looks so. I should not have merged a PR with the RC reference, but we can quickly integrate the stuff before the weekly

**Comment by oleg-nenashev** [comment]
If we cannot finish the process by Friday EoD, I will submit a revert PR

**Comment by kuisathaverat** [comment]
@oleg-nenashev I've made the changes on the core https://github.com/jenkinsci/jenkins/pull/4109 and I just release the trilead-api plugin

**Pull request merged**
